### ISSUE TITLE
Skip API key requirement when modelType is 'ollama'

### ISSUE THREAD

**Issue by elcoosp** [kmbro/langextract-typescript#1]
### GitHub Issue: Skip `apiKey` Requirement for Ollama Models + Improve Type Safety

---

**Title**  
Enforce required `apiKey` for cloud models while making it optional for Ollama + improve type safety

**Description**  
We need to fix two related issues:
1. **Runtime validation**: The `apiKey` is currently required even for local Ollama models (`modelType === 'ollama'`), forcing users to pass dummy values
2. **Type safety**: The current `options` type allows omitting `apiKey` for cloud-hosted models (like Gemini), which should be required

**Steps to Reproduce**  
1. Run extraction with Ollama without API key:
```typescript
extract(rawSpec, {
  modelType: 'ollama',
  // No apiKey provided
});
// Throws error even though Ollama doesn't need API key
```

2. Run extraction with Gemini without API key (type system doesn't catch):
```typescript
extract(rawSpec, {
  modelType: 'gemini',
  // No apiKey provided (TS doesn't error but runtime fails)
});
// TypeScript doesn't flag this as error, but runtime fails
```

**Expected Behavior**  
- `apiKey` should be optional when `modelType === 'ollama'`
- `apiKey` should be required for cloud models (Gemini, etc.)
- Type system should enforce these requirements

**Proposed Solution**  

**1. Runtime Fix (src/index.ts):**
```typescript
// Before
if (!apiKey) {
  throw new Error("API key must be provided...");
}

// After
if (modelType !== 'ollama' && !apiKey) {
  throw new Error("API key must be provided for cloud-hosted models...");
}
```

**2. Type Definition Fix:**  
Implement discriminated union types:
```typescript
type BaseExtractOptions = {
  promptDescription?: string;
  examples?: ExampleData[];
  modelId?: string;
  formatType?: FormatType;
  // ...other common options...
};

type CloudModelOptions = BaseExtractOptions & {
  modelType: 'gemini' | 'openai'; // Add other cloud models
  apiKey: string; // Required for cloud models
};

type OllamaModelOptions = BaseExtractOptions & {
  modelType: 'ollama';
  apiKey?: never; // Prohibits apiKey for Ollama
  modelUrl: string; // Make modelUrl required for Ollama
};

type ExtractOptions = CloudModelOptions | OllamaModelOptions;
```

**3. Update Function Signature:**
```typescript
async function extract(
  textOrDocuments: string | Document[],
  options: ExtractOptions
) {
  // Implementation
}
```

**Benefits**  
1. Eliminates need for dummy API keys with Ollama
2. Type system prevents:
   - Omitting `apiKey` for cloud models
   - Passing `apiKey` with Ollama (which could cause confusion)
   - Omitting `modelUrl` for Ollama (which is required)
3. Clearer developer experience with model-specific requirements

**Affected Files**  
1. [`src/index.ts`](https://github.com/.../src/index.ts) (runtime validation)
2. [`src/index.ts`](https://github.com/kmbro/langextract-typescript/blob/adb2a8c53413d336faa26c812ff9f5262d461bb5/src/index.ts#L64) (type definitions)

**Comment by kmbro** [comment]
Hi! Feel free to open a PR to fix this. I think it sounds reasonable.
